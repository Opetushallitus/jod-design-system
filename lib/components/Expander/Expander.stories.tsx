import type { Meta, StoryObj } from '@storybook/react';

import { Expander } from './Expander';

const meta = {
  title: 'Primitives/Expander',
  component: Expander,
  tags: ['autodocs'],
} satisfies Meta<typeof Expander>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=542%3A7857',
  },
};

export const Default: Story = {
  parameters: {
    ...parameters,
  },
  args: {
    label: 'Label',
    description: 'Description',
    children: <div>Children here</div>,
  },
};
