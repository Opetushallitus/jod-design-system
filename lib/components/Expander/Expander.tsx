import { Disclosure, DisclosureButton, DisclosurePanel } from '@headlessui/react';
import React from 'react';

export interface ExpanderProps {
  label: string;
  description: string;
  children?: React.ReactNode;
}

export const Expander = ({ label, description, children }: ExpanderProps) => {
  const labelId = React.useId();

  return (
    <Disclosure as="div" className="rounded-[20px] border-[3px] border-[#767676] p-4">
      {({ open }) => (
        <>
          <DisclosureButton className="flex w-full flex-col">
            <div className="mb-3 flex w-full flex-row justify-between">
              <span id={labelId} className="text-[20px] font-bold text-[#333333]">
                {label}
              </span>
              <span className="material-symbols-outlined size-32 select-none text-secondary-gray">
                {open ? 'keyboard_arrow_up' : 'keyboard_arrow_down'}
              </span>
            </div>
            <span className="mb-4 text-start text-[12px] font-bold text-[#767676]">{description}</span>
          </DisclosureButton>
          <DisclosurePanel role="region" aria-labelledby={labelId}>
            {children}
          </DisclosurePanel>
        </>
      )}
    </Disclosure>
  );
};
