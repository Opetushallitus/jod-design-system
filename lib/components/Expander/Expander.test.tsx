import { render, screen, cleanup } from '@testing-library/react';
import { afterEach, describe, expect, test } from 'vitest';
import '@testing-library/jest-dom/vitest';
import userEvent from '@testing-library/user-event';

import { Expander } from './Expander';

afterEach(() => {
  cleanup();
});

describe('Snapshot testing', () => {
  test('Base button', () => {
    const { container } = render(
      <Expander label="Label" description="Description">
        content
      </Expander>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Expander', () => {
  test('renders label and description correctly', () => {
    const label = 'Test Label';
    const description = 'Test Description';
    render(<Expander label={label} description={description} />);

    const labelElement = screen.getByText(label);
    const descriptionElement = screen.getByText(description);

    expect(labelElement).toBeInTheDocument();
    expect(descriptionElement).toBeInTheDocument();
  });

  test('expands and collapses on button click', async () => {
    const user = userEvent.setup();
    const label = 'Test Label';
    const description = 'Test Description';
    render(<Expander label={label} description={description} />);

    const button = screen.getByRole('button');
    expect(screen.queryByRole('region')).toBeNull();

    await user.click(button);
    expect(screen.getByRole('region')).not.toBeNull();
    await user.click(button);

    expect(screen.queryByRole('region')).toBeNull();
  });
});
