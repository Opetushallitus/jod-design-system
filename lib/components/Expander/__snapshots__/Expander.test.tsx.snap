// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing > Base button 1`] = `
<div
  class="rounded-[20px] border-[3px] border-[#767676] p-4"
  data-headlessui-state=""
>
  <button
    aria-expanded="false"
    class="flex w-full flex-col"
    data-headlessui-state=""
    id="headlessui-disclosure-button-:r1:"
    type="button"
  >
    <div
      class="mb-3 flex w-full flex-row justify-between"
    >
      <span
        class="text-[20px] font-bold text-[#333333]"
        id=":r0:"
      >
        Label
      </span>
      <span
        class="material-symbols-outlined size-32 select-none text-secondary-gray"
      >
        keyboard_arrow_down
      </span>
    </div>
    <span
      class="mb-4 text-start text-[12px] font-bold text-[#767676]"
    >
      Description
    </span>
  </button>
</div>
`;
